implement GSSAPI MIC token

bodgit opened the pull request:
This hopefully is the last missing piece that finally fixes #65. Based heavily on the Wrap token support added in PR #83.

I've not documented or added tests yet as currently the intended use for the code doesn't work compared to another existing GSSAPI implementation and I'm not entirely sure if it's my implementation/addition or something wrong with how I'm using the rest of this library. Rather than clutter this PR, I'll keep the discussion in #65 to hopefully work out what's going wrong.

jcmturner commented on gssapi/MICToken.go:
```diff
@@ -0,0 +1,195 @@
+}
+
+// Build a header suitable for a checksum computation
+func getMICChecksumHeader(flags byte, senderSeqNum uint64) []byte {
```
I think this might be neater as a method on the MICToken: ``func (mt *MICToken) checksumHeader() []byte``

This way you don't need to pass in the flags and sequence number as arguments.

jcmturner commented on gssapi/MICToken.go:
```diff
@@ -0,0 +1,195 @@
+// and will check the according flag, returning an error if the token does not match the expectation.
+func (mt *MICToken) Unmarshal(b []byte, expectFromAcceptor bool) error {
+	if len(b) < MICHdrLen {
+		return errors.New("bytes shorter than hedaer length")
```
typo ``hedaer``

jcmturner commented on gssapi/MICToken.go:
```diff
@@ -0,0 +1,195 @@
+// Other flags are set to 0.
+// Note that in certain circumstances you may need to provide a sequence number that has been defined earlier.
+// This is currently not supported.
+func NewInitiatorMICToken(payload []byte, key types.EncryptionKey) (*MICToken, error) {
```
What is the reason to return a pointer here?

bodgit replied:
Purely because that was what the WrapToken implementation did.

jcmturner commented on gssapi/MICToken.go:
```diff
@@ -0,0 +1,195 @@
+			hex.EncodeToString(b[0:2]))
+	}
+	flags := b[2]
+	isFromAcceptor := flags&0x01 == 1
```
Would be good to create constants for the flag values and use the constant in the comparison.
The values are here: https://tools.ietf.org/html/rfc4121#section-4.2.2
So I suggest:
```
const (
    MICTokenFlagSentByAcceptor = 0
    MICTokenFlagSealed = 1
    MICTokenFlagAcceptorSubkey = 2
)
```
therefore...
```
isFromAcceptor := flags&0x01 == MICTokenFlagSealed
```

jcmturner commented:
I'll take a look as I can see merit in the idea. However this would likely need a new major release as it would probably change public functions and methods. I think we should get the MICToken working with test coverage and then we can look to refactor GSSAPI tokens to have a consistent interface. This way we deliver the functionality and then we can optimise the code once we know what the working and tested solutions are.

bodgit commented:
I've now added tests and I think I've addressed your comments, bar the one about returning a struct pointer. Up to you whether you favour correctness or consistency between both token implementations.

My only question now is having added the tests, the payload is treated differently in MIC tokens as it's not `Marshal()`'d and encoded in the token, it's usually sent separately. Does it therefore make sense to keep it part of the struct or just pass it in separately as an argument to the `NewInitiatorMICToken()`, `SetChecksum()` and `VerifyChecksum()` methods? The disadvantage I suppose is it makes the method signatures different between both token types.

bodgit commented on gssapi/MICToken.go:
```diff
@@ -0,0 +1,202 @@
+)
+
+const (
+	hdrLen = 16 // Length of the MIC Token's header
```
Fixed.

bodgit commented on gssapi/MICToken.go:
```diff
@@ -0,0 +1,202 @@
+}
+
+// Return the 2 bytes identifying a GSS API MIC token
+func getGSSMICTokenId() *[2]byte {
```
Fixed.

jcmturner commented:
@bodgit I've been working on a refactor of the GSS-API and SPNEGO in the gssapi branch: https://github.com/jcmturner/gokrb5/tree/gssapi

Would be interested in any thoughts you had...

bodgit commented:
@jcmturner I will definitely take a look if it's going to have an affect on the API. I'm guessing some of the functions/methods have changed?

All of my code that calls into your library is here: https://github.com/bodgit/tsig/blob/master/gss/gokrb5.go

jcmturner commented:
I've merged the gssapi branch into the branch being used to draft v7 of gokrb5: https://github.com/jcmturner/gokrb5/tree/draft-v7

bodgit commented:
I think I've adjusted my code to work with your changes, I have this compile-tested commit: https://github.com/bodgit/tsig/commit/fbee830882a85285570c4a315eb3e098911e8ecb

The only bit I'm not 100% on is with the `gssapi.MechToken` -> `spnego.KRB5Token` changes, `spnego.NewKRB5TokenAPREQ()` wants a `service.Settings` which I've satisfied by using `service.NewSettings(nil)`, is that correct?

Other than that, most of the other changes were easy to figure out.

jcmturner commented:
Hi @bodgit thanks for trying this out. I'm looking into the serivce.Settings as this may not be needed on the client side. Therefore I think I may be able to remove the need for it on this function. I'll test that and if so merge an update into the draft-v7 branch.
